flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java

```diff
@@ -0,0 +1,408 @@
+
+	/**
+	 * Valves that control how watermarks and stream statuses from the 2 inputs are forwarded.
+	 */
+	private final StatusWatermarkValve statusWatermarkValve1;
+	private final StatusWatermarkValve statusWatermarkValve2;
+
+	/**
+	 * Stream status for the two inputs. We need to keep track for determining when
+	 * to forward stream status changes downstream.
+	 */
+	private StreamStatus firstStatus;
+	private StreamStatus secondStatus;
+
+	private int availableInputsMask;
+
+	private int lastReadInputIndex;
+
+	private InputSelection inputSelection;
+
+	private Counter numRecordsIn;
+
+	private boolean initialized;
+
+	public StreamTwoInputSelectableProcessor(
+		Collection<InputGate> inputGates1,
+		Collection<InputGate> inputGates2,
+		TypeSerializer<IN1> inputSerializer1,
+		TypeSerializer<IN2> inputSerializer2,
+		Object lock,
+		IOManager ioManager,
+		StreamStatusMaintainer streamStatusMaintainer,
+		TwoInputStreamOperator<IN1, IN2, ?> streamOperator,
+		WatermarkGauge input1WatermarkGauge,
+		WatermarkGauge input2WatermarkGauge) {
+
+		checkState(streamOperator instanceof InputSelectable);
+
+		this.streamOperator = checkNotNull(streamOperator);
+		this.inputSelector = (InputSelectable) streamOperator;
+
+		this.lock = checkNotNull(lock);
+
+		InputGate unionedInputGate1 = InputGateUtil.createInputGate(inputGates1.toArray(new InputGate[0]));
+		InputGate unionedInputGate2 = InputGateUtil.createInputGate(inputGates2.toArray(new InputGate[0]));
+
+		// create a Input instance for each input
+		this.input1 = new StreamTaskNetworkInput(new BarrierDiscarder(unionedInputGate1), inputSerializer1, ioManager, 0);
+		this.input2 = new StreamTaskNetworkInput(new BarrierDiscarder(unionedInputGate2), inputSerializer2, ioManager, 1);
+
+		this.statusWatermarkValve1 = new StatusWatermarkValve(
+			unionedInputGate1.getNumberOfInputChannels(),
+			new ForwardingValveOutputHandler(streamOperator, lock, streamStatusMaintainer, input1WatermarkGauge, 0));
+		this.statusWatermarkValve2 = new StatusWatermarkValve(
+			unionedInputGate2.getNumberOfInputChannels(),
+			new ForwardingValveOutputHandler(streamOperator, lock, streamStatusMaintainer, input2WatermarkGauge, 1));
+
+		this.firstStatus = StreamStatus.ACTIVE;
+		this.secondStatus = StreamStatus.ACTIVE;
+
+		this.availableInputsMask = (int) new InputSelection.Builder().select(1).select(2).build().getInputMask();
+
+		this.lastReadInputIndex = 1; // always try to read from the first input
+
+		this.initialized = false;
+
+	}
+
+	public boolean processInput() throws Exception {
+		if (!initialized) {
+			initialize();
+		}
+
+		int readingInputIndex = inputSelection.fairSelectNextIndexOutOf2(availableInputsMask, lastReadInputIndex);
+		if (readingInputIndex == -1) {
+			return waitForAvailableInput(inputSelection);
+		}
+		lastReadInputIndex = readingInputIndex;
+
+		// if the input selection is ALL and availableInputsMask is not ALL,
+		// always try to check and set the availability of another input
```

Comment:
can you add to the comment:
```
// to avoid starvation, if the input selection is ALL and availableInputsMask is not ALL,
// always try to check and set the availability of another input
// TODO: because this can be a costly operation (checking volatile inside CompletableFuture` this might be optimized to only check once per processed NetworkBuffer
```
?